multicast_expert/interfaces.py

```diff
@@ -0,0 +1,353 @@
+                # Found a match!
+                return [iface]
+
+    if isinstance(specifier, str):
+        try:
+            specifier_ip_addr = ipaddress.ip_address(specifier)
+        except Exception as ex:
+            message = f"Specifier '{specifier}' does not appear to be a valid interface name or IP address!"
+            raise MulticastExpertError(message) from ex
+    else:
+        specifier_ip_addr = specifier
+
+    is_ipv6 = isinstance(specifier_ip_addr, IPv6Address)
+
+    result = []
+    for iface in ifaces:
+        # Annoyingly IPv[4/6]Network does not compare as equal to IPv[4/6]Address, so we have to convert
+        addrs_to_check: set[IPv4Address | IPv6Address] = set()
+        if is_ipv6:
+            for addr in iface.ip6_addrs:
+                # go through string to work around https://github.com/python/cpython/issues/129538
+                addr_string = ip_interface_to_ip_string(addr)
+                if cast(IPv6Address, specifier_ip_addr).scope_id is None:
+                    # Trim off the scope ID from the address string
+                    addr_string = addr_string.split("%")[0]
+                addrs_to_check.add(IPv6Address(addr_string))
+        else:
+            addrs_to_check.update(addr.ip for addr in iface.ip4_addrs)
+
+        if specifier_ip_addr in addrs_to_check:
+            result.append(iface)
+
+    if len(result) == 0:
+        message = f"No matches found for interface IP address {specifier_ip_addr!s}"
+        raise MulticastExpertError(message)
+
+    return result
+
+
+def find_interfaces(
+    specifiers: Iterable[IfaceSpecifier], *, ifaces: Sequence[IfaceInfo] | None = None
+) -> list[IfaceInfo]:
+    """
+    Find interfaces (represented by IfaceInfo objects) based on a collection of interface specifiers.
+
+    If no interfaces match any individual specifier, a MulticastExpertError is raised.
+
+    :param specifiers: Specifier for each interface you want to find. If a specifier is an IfaceInfo already,
+        it will simply be added to the result list.
+    :param ifaces: If set, these interfaces will be searched using the specifier.
+        If not set, then the current set of interfaces will be scanned from the machine.
+
+    :return: Found interface(s). This function will usually return as many IfaceInfos as the number of
+        specifiers you passed in. However, if one IP address is used on multiple interfaces, and you
+        pass in that IP address as a specifier, multiple interfaces will be matched for that specifier. Also, if
+        multiple specifiers matched the same interface, the results will be deduplicated.
+    """
+    if ifaces is None:
+        ifaces = scan_interfaces()
+    results = [iface for specifier in specifiers for iface in _find_interfaces_for_specifier(specifier, ifaces=ifaces)]
+    result_dict = {iface.index: iface for iface in results}  # deduplicate interfaces
+    return list(result_dict.values())
```

Comment:
Hmm... It turns out that this will scan interfaces if `ifaces` is `None`, even when all specifiers are `IfaceInfo`s. I think this should be addressed here (as opposed to in `McastR|TxSocket.__init__`), with something like this:

```suggestion
    specifiers = list(specifiers)
    if all(isinstance(x, IfaceInfo) for x in specifiers):
        results = cast(list[IfaceInfo], specifiers)
    else:
        if ifaces is None:
            ifaces = scan_interfaces()
        results = [
            iface for specifier in specifiers for iface in _find_interfaces_for_specifier(specifier, ifaces=ifaces)
        ]
```

Reply:
Ooh good catch. I'll add a test too!